src/k8s/pkg/k8sd/api/certificates_refresh.go

```diff
@@ -60,8 +60,97 @@ func (e *Endpoints) postRefreshCertsRun(s state.State, r *http.Request) response
 	if isWorker {
 		return refreshCertsRunWorker(s, r, snap)
 	}
-	// TODO: Control Plane refresh
-	return response.InternalError(fmt.Errorf("not implemented yet"))
+	return refreshCertsRunControlPlane(s, r, snap)
+}
+
+// refreshCertsRunControlPlane refreshes the certificates for a control plane node.
+func refreshCertsRunControlPlane(s state.State, r *http.Request, snap snap.Snap) response.Response {
+	req := apiv1.RefreshCertificatesRunRequest{}
+	if err := utils.NewStrictJSONDecoder(r.Body).Decode(&req); err != nil {
+		return response.BadRequest(fmt.Errorf("failed to parse request: %w", err))
+	}
+
+	clusterConfig, err := databaseutil.GetClusterConfig(r.Context(), s)
+	if err != nil {
+		return response.InternalError(fmt.Errorf("failed to recover cluster config: %w", err))
+	}
+
+	nodeIP := net.ParseIP(s.Address().Hostname())
+	if nodeIP == nil {
+		return response.InternalError(fmt.Errorf("failed to parse node IP address %q", s.Address().Hostname()))
+	}
+
+	serviceIPs, err := utils.GetKubernetesServiceIPsFromServiceCIDRs(clusterConfig.Network.GetServiceCIDR())
+	if err != nil {
+		return response.InternalError(fmt.Errorf("failed to get IP address(es) from ServiceCIDR %q: %w", clusterConfig.Network.GetServiceCIDR(), err))
+	}
+
+	extraIPs, extraNames := utils.SplitIPAndDNSSANs(req.ExtraSANs)
+
+	certificates := pki.NewControlPlanePKI(pki.ControlPlanePKIOpts{
+		Hostname:                  s.Name(),
+		IPSANs:                    append(append([]net.IP{nodeIP}, serviceIPs...), extraIPs...),
+		Seconds:                   req.ExpirationSeconds,
+		DNSSANs:                   extraNames,
+		AllowSelfSignedCA:         true,
+		IncludeMachineAddressSANs: true,
+	})
+
+	certificates.CACert = clusterConfig.Certificates.GetCACert()
+	certificates.CAKey = clusterConfig.Certificates.GetCAKey()
+	certificates.ClientCACert = clusterConfig.Certificates.GetClientCACert()
+	certificates.ClientCAKey = clusterConfig.Certificates.GetClientCAKey()
+	certificates.FrontProxyCACert = clusterConfig.Certificates.GetFrontProxyCACert()
+	certificates.FrontProxyCAKey = clusterConfig.Certificates.GetFrontProxyCAKey()
+	certificates.K8sdPrivateKey = clusterConfig.Certificates.GetK8sdPrivateKey()
+	certificates.K8sdPublicKey = clusterConfig.Certificates.GetK8sdPublicKey()
+	certificates.ServiceAccountKey = clusterConfig.Certificates.GetServiceAccountKey()
+
+	if err := certificates.CompleteCertificates(); err != nil {
+		return response.InternalError(fmt.Errorf("failed to generate new control plane certificates: %w", err))
+	}
+
+	if _, err := setup.EnsureControlPlanePKI(snap, certificates); err != nil {
+		return response.InternalError(fmt.Errorf("failed to write control plane certificates: %w", err))
+	}
+
+	// Generate kubeconfigs
+	for _, kubeconfig := range []struct {
```

Comment:
Let's use the existing util for this: https://github.com/canonical/k8s-snap/blob/67f8883a2fc929dcb80c8d7ed0b5437ec90b3e42/src/k8s/pkg/k8sd/app/cluster_util.go#L15

Or is there a reason not to?

Reply:
This will create an import cycle. We can work around this by creating a new utils package, which will contain the utilities for creating the kubeconfigs.

Reply:
you could move this to `setup` package